marathon/src/main/scala/io/buoyant/marathon/v2/Api.scala

```diff
@@ -51,16 +57,21 @@ object Api {
       case http.Status.Ok =>
         val addrs = readJson[AppRsp](rsp.content).map(_.toAddresses)
         Future.const(addrs)
-      case _ =>
-        Future.exception(UnexpectedResponse(rsp))
+      case http.Status.Unauthorized =>
+        Future.exception(UnauthorizedResponse(rsp))
+      case _ => Future.exception(UnexpectedResponse(rsp))
     }
 
+  private[v2] case class AuthToken(token: Option[String])
+
+  private[v2] case class UnauthorizedResponse(rsp: http.Response) extends Throwable
+
   private[this] case class UnexpectedResponse(rsp: http.Response) extends Throwable
 
   private[this] val mapper = new ObjectMapper with ScalaObjectMapper
   mapper.registerModule(DefaultScalaModule)
   mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
-  private[this] def readJson[T: Manifest](buf: Buf): Try[T] = {
+  def readJson[T: Manifest](buf: Buf): Try[T] = {
```

Comment:
making `readJson` public makes me think this could be useful across the broader codebase. thoughts about moving it out into some utility code?